watcher_events.go

```diff
@@ -139,7 +139,46 @@ func handleWatchEvent(ctx context.Context, event *watch.Event, cutoffTime metav1
 		logger.Error().Msgf("Cannot get Sentry hub from context")
 		return
 	}
+
+	// To avoid concurrency issue
+	hub = hub.Clone()
 	hub.WithScope(func(scope *sentry.Scope) {
+
+		// Find the object meta that the event is about
+		objectMeta, err := findObjectMeta(ctx, eventObject.InvolvedObject.Kind, eventObject.InvolvedObject.Namespace, eventObject.InvolvedObject.Name)
+		if err != nil {
+			return
+		}
+
+		// Search for alternative DSN in the annotations
+		altDsn, err := searchDsn(ctx, *objectMeta)
+		if err != nil {
+			return
+		}
+
+		// If we did find an alternative DSN
+		if altDsn != "" {
+			// Attempt to retrieve the corresponding client
+			client, err := dsnData.GetClient(altDsn)
+			if err != nil {
+				return
+			}
+
+			if client == nil {
+				newOptions := hub.Client().Options()
+				newOptions.Dsn = altDsn
+				client, err = dsnData.AddClient(newOptions)
+				if err != nil {
+					return
+				}
+			}
+
+			// Bind the alternative client to the top layer
+			hub.BindClient(client)
+		}
```

Comment:
This block is basically duplicated three times across different files/types of events (events, pods, crons).
Please think of a way to reuse the code.